help.go

```diff
@@ -7,130 +7,154 @@ import (
 	"io"
 	"reflect"
 	"strings"
-
-	"github.com/aymerick/raymond"
 )
 
 const (
-	defaultIndent   = 2
-	defaultTemplate = `
-{{#with App}}
-usage: {{Name}}
-
-{{#wrap}}
-{{Help}}
-{{/wrap}}
-
-Flags:
-{{#indent}}
-{{formatFlags Flags}}
-{{/indent}}
-
-{{#if Children}}
-{{#indent}}
-{{#each Children}}
-{{Name}}
-{{/each}}
-{{/indent}}
-{{/if}}
-{{/with}}
-`
+	defaultIndent = 2
 )
 
-var defaultHelpTemplate = raymond.MustParse(strings.TrimSpace(defaultTemplate))
+func PrintHelp(ctx *Context) error {
+	w := newHelpWriter(guessWidth(ctx.App.Stdout))
+	selected := ctx.Selected()
+	if selected == nil {
+		printApp(w, ctx.App.Application)
+	} else {
+		printCommand(w, ctx.App.Application, selected)
+	}
+	return w.Write(ctx.App.Stdout)
+}
 
-func init() {
-	defaultHelpTemplate.RegisterHelpers(map[string]interface{}{
-		"indent": func(options *raymond.Options) string {
-			indent, ok := options.HashProp("depth").(int)
-			if !ok {
-				indent = 2
-			}
-			width := options.Data("width").(int)
-			frame := options.NewDataFrame()
-			frame.Set("width", width-indent)
-			indentStr := strings.Repeat(" ", indent)
-			lines := strings.Split(options.FnData(frame), "\n")
-			for i, line := range lines {
-				lines[i] = indentStr + line
-			}
-			return strings.Join(lines, "\n")
-		},
-		"formatFlags": func(flags []*Flag, options *raymond.Options) string {
-			rows := [][2]string{}
-			haveShort := false
-			for _, flag := range flags {
-				if flag.Short != 0 {
-					haveShort = true
-					break
-				}
-			}
-			for _, flag := range flags {
-				if !flag.Hidden {
-					rows = append(rows, [2]string{formatFlag(haveShort, flag), flag.Help})
-				}
+func printApp(w *helpWriter, app *Application) {
+	w.Printf("usage: %s", app.Summary())
+	printNodeDetail(w, &app.Node)
+}
+
+func printCommand(w *helpWriter, app *Application, cmd *Command) {
+	w.Printf("usage: %s %s", app.Name, cmd.Summary())
+	printNodeDetail(w, cmd)
+}
+
+func printNodeDetail(w *helpWriter, node *Node) {
+	if node.Help != "" {
+		w.Print("")
+		w.Wrap(node.Help)
+	}
+	if len(node.Flags) > 0 {
+		w.Printf("")
+		w.Printf("Flags:")
+		writeFlags(w.Indent(), node.Flags)
+	}
+	cmds := node.Leaves()
+	if len(cmds) > 0 {
```

Comment:
Reduce indentation with this?

```go
if len(cmds) == 0 {
    return;
}
```

Reply:
Normally I would agree, but I preferred it like this so that it's visually equivalent to the other two blocks in the function. That is, each `if` block deals with a single component of the help.

Reply:
Fair enough!